File: runtime-bevy/crates/threenative_runtime/src/runtime_gameplay_host.rs
use serde_json::{json, Value};
use threenative_loader::{LoadedBundle, SystemIr};

use crate::systems_effects::{
    apply_system_effects, NativeSystemCommandEffect, NativeSystemEffects,
    NativeSystemResourceEffect,
};

pub fn trace_runtime_gameplay_host(bundle: &mut LoadedBundle) -> Value {
    let timer_system = system_by_name(bundle, "timerAndObserver");
    let spawn_system = system_by_name(bundle, "spawnRenderable");
    let remove_system = system_by_name(bundle, "removeRenderable");
    let mut command_flush = Vec::new();
    let mut local_values = vec![read_local_counter(bundle)];

    let timer = apply_system_effects(bundle, &timer_system, &timer_effects(), 1, 1)
        .expect("timer effects are declared");
    command_flush.extend(timer.entries.iter().map(entry_key));
    local_values.push(read_local_counter(bundle));

    let spawn = apply_system_effects(bundle, &spawn_system, &spawn_effects(), 2, 2)
        .expect("spawn effects are declared");
    command_flush.extend(spawn.entries.iter().map(entry_key));
    let spawned_renderer_handles = renderer_handles(bundle);
    local_values.push(read_local_counter(bundle));

    let teardown_order = command_flush.len() + 1;
    let remove = apply_system_effects(bundle, &remove_system, &remove_effects(), 3, 3)
        .expect("remove effects are declared");
    command_flush.extend(remove.entries.iter().map(entry_key));

    json!({
        "async": {
            "channels": channels(bundle),
            "timers": timers(bundle)
        },
        "boundaries": [
            boundary("TN_RUNTIME_DYNAMIC_PLUGIN_UNSUPPORTED", "Keep runtime plugins outside portable bundles or promote a bounded SDK declaration."),
            boundary("TN_RUNTIME_RAW_HANDLE_UNSUPPORTED", "Use portable entity, component, asset, and service identifiers instead of backend handles."),
            boundary("TN_RUNTIME_UNBOUNDED_ASYNC_UNSUPPORTED", "Use fixed-tick timers, fixed-trace channels, or target-specific adapters.")
        ],
        "diagnostics": diagnostics(),
        "eventWindows": [
            { "event": "DamageEvent", "framesVisible": [1, 2], "policy": "clear-after-post-update" },
            { "event": "Spawned", "framesVisible": [2, 3], "policy": "clear-after-post-update" },
            { "event": "TimerElapsed", "framesVisible": [1], "policy": "clear-after-post-update" }
        ],
        "hooks": [
            { "component": "Health", "entity": "runtime.enemy", "hook": "onAdd", "order": 1 },
            { "component": "Health", "entity": "runtime.enemy", "hook": "onInsert", "order": 2 },
            { "component": "Health", "entity": "runtime.enemy", "hook": "onRemove", "order": 3 }
        ],
        "lifecycle": {
            "appState": [
                { "from": "boot", "schedule": "update", "to": "playing" },
                { "from": "playing", "schedule": "postUpdate", "to": "settled" }
            ],
            "commandFlush": command_flush,
            "localState": [{ "key": "LocalCounter.value", "resetOnTeardown": true, "values": local_values }]
        },
        "observers": [{ "event": "DamageEvent", "route": ["target:player.weapon"], "status": "stopped" }],
        "reconciliation": {
            "finalRendererHandles": renderer_handles(bundle),
            "rendererTeardown": [{ "entity": "runtime.enemy", "order": teardown_order, "rendererHandle": "renderer:runtime.enemy", "removed": true }],
            "spawnedRendererHandles": spawned_renderer_handles
        },
        "schema": "threenative.runtime-gameplay-host",
        "version": "0.1.0"
    })
}

fn timer_effects() -> NativeSystemEffects {
    NativeSystemEffects {
        commands: vec![NativeSystemCommandEffect {
            command: "emitEvent".to_owned(),
            component: None,
            components: None,
            entity: Some(String::new()),
            event: Some("TimerElapsed".to_owned()),
            payload: Some(json!({ "id": "boundedTimer", "tick": 1 })),
            value: None,
            ..Default::default()
        }],
        events: Vec::new(),
        patches: Vec::new(),
        resources: vec![NativeSystemResourceEffect {
            resource: "LocalCounter".to_owned(),
            value: json!({ "value": 1 }),
        }],
        services: Vec::new(),
    }
}

fn spawn_effects() -> NativeSystemEffects {
    NativeSystemEffects {
        commands: vec![
            NativeSystemCommandEffect {
                command: "spawn".to_owned(),
                component: None,
                components: Some(json!({
                    "Health": { "current": 4, "max": 4 },
                    "MeshRenderer": { "material": "mat.spawned", "mesh": "primitive.box" },
                    "Transform": { "position": [2, 0, 0] }
                })),
                entity: Some("runtime.enemy".to_owned()),
                event: None,
                payload: None,
                value: None,
                ..Default::default()
            },
            NativeSystemCommandEffect {
                command: "emitEvent".to_owned(),
                component: None,
                components: None,
                entity: Some(String::new()),
                event: Some("Spawned".to_owned()),
                payload: Some(json!({ "entity": "runtime.enemy" })),
                value: None,
                ..Default::default()
            },
        ],
        events: Vec::new(),
        patches: Vec::new(),
        resources: vec![
            NativeSystemResourceEffect {
                resource: "GameState".to_owned(),
                value: json!({ "combat": "engaged", "phase": "playing" }),
            },
            NativeSystemResourceEffect {
                resource: "LocalCounter".to_owned(),
                value: json!({ "value": 2 }),
            },
        ],
        services: Vec::new(),
    }
}

fn remove_effects() -> NativeSystemEffects {
    NativeSystemEffects {
        commands: vec![
            NativeSystemCommandEffect {
                command: "removeComponent".to_owned(),
                component: Some("Health".to_owned()),
                components: None,
                entity: Some("runtime.enemy".to_owned()),
                event: None,
                payload: None,
                value: None,
                ..Default::default()
            },
            NativeSystemCommandEffect {
                command: "despawn".to_owned(),
                component: None,
                components: None,
                entity: Some("runtime.enemy".to_owned()),
                event: None,
                payload: None,
                value: None,
                ..Default::default()
            },
        ],
        events: Vec::new(),
        patches: Vec::new(),
        resources: vec![NativeSystemResourceEffect {
            resource: "GameState".to_owned(),
            value: json!({ "combat": "safe", "phase": "settled" }),
        }],
        services: Vec::new(),
    }
}

fn system_by_name(bundle: &LoadedBundle, name: &str) -> SystemIr {
    bundle
        .systems
        .as_ref()
        .and_then(|systems| systems.systems.iter().find(|system| system.name == name))
        .cloned()
        .unwrap_or_else(|| panic!("missing runtime gameplay host system '{name}'"))
}

fn renderer_handles(bundle: &LoadedBundle) -> Vec<String> {
    let mut handles = bundle
        .world
        .entities
        .iter()
        .filter(|entity| entity.components.mesh_renderer.is_some())
        .map(|entity| format!("renderer:{}", entity.id))
        .collect::<Vec<_>>();
    handles.sort();
    handles
}

fn read_local_counter(bundle: &LoadedBundle) -> i64 {
    bundle
        .world
        .resources
        .get("LocalCounter")
        .and_then(|value| value.get("value"))
        .and_then(Value::as_i64)
        .unwrap_or(0)
}

fn channels(bundle: &LoadedBundle) -> Value {
    Value::Array(
        bundle
            .systems
            .as_ref()
            .map(|systems| {
                systems
                    .channels
                    .iter()
                    .map(|channel| {
                        json!({
                            "delivery": channel.delivery,
                            "event": channel.event,
                            "id": channel.id,
                            "status": "ready"
                        })
                    })
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default(),
    )
}

fn timers(bundle: &LoadedBundle) -> Value {
    Value::Array(bundle.systems.as_ref().map(|systems| {
        systems.tasks.iter().map(|task| {
            let event = task.channel.as_ref()
                .and_then(|channel_id| systems.channels.iter().find(|channel| channel.id == *channel_id))
                .map(|channel| channel.event.clone())
                .unwrap_or_default();
            json!({ "event": event, "firedAtTick": 1, "id": task.id, "mode": task.mode, "status": "fired" })
        }).collect::<Vec<_>>()
    }).unwrap_or_default())
}

fn diagnostics() -> Value {
    json!([
        diagnostic(
            "TN_UNSUPPORTED_FEATURE_PROMISE",
            "systems.ir.json/runtime/unsupportedFeatures/promise"
        ),
        diagnostic(
            "TN_UNSUPPORTED_FEATURE_RAW_RUNTIME_HANDLE",
            "systems.ir.json/runtime/unsupportedFeatures/rawRuntimeHandle"
        ),
        diagnostic(
            "TN_UNSUPPORTED_FEATURE_RUNTIME_PLUGIN",
            "systems.ir.json/runtime/unsupportedFeatures/runtimePlugin"
        ),
        diagnostic(
            "TN_UNSUPPORTED_FEATURE_TIMER",
            "systems.ir.json/runtime/unsupportedFeatures/timer"
        ),
        diagnostic(
            "TN_UNSUPPORTED_FEATURE_WORKER",
            "systems.ir.json/runtime/unsupportedFeatures/worker"
        )
    ])
}

fn diagnostic(code: &str, path: &str) -> Value {
    json!({
        "code": code,
        "message": unsupported_message(code),
        "path": path,
        "severity": "error",
        "suggestion": "Remove the unsupported declaration or replace it with a portable SDK/IR declaration."
    })
}

fn unsupported_message(code: &str) -> String {
    let feature = code
        .trim_start_matches("TN_UNSUPPORTED_FEATURE_")
        .to_lowercase();
    format!(
        "Feature '{}' is outside the portable runtime scope.",
        feature.to_camel_case()
    )
}

trait CamelCase {
    fn to_camel_case(&self) -> String;
}

impl CamelCase for str {
    fn to_camel_case(&self) -> String {
        let mut output = String::new();
        let mut uppercase_next = false;
        for character in self.chars() {
            if character == '_' {
                uppercase_next = true;
            } else if uppercase_next {
                output.extend(character.to_uppercase());
                uppercase_next = false;
            } else {
                output.push(character);
            }
        }
        output
    }
}

fn boundary(code: &str, suggestion: &str) -> Value {
    json!({ "code": code, "status": "diagnostic-only", "suggestion": suggestion })
}

fn entry_key(entry: &crate::systems_effects::NativeSystemEffectLogEntry) -> String {
    format!(
        "{}:{}:{}:{}:{}:{}:{}",
        entry.schedule,
        entry.system,
        entry.kind,
        entry.command.as_deref().unwrap_or(""),
        entry.entity.as_deref().unwrap_or(""),
        entry.component.as_deref().unwrap_or(""),
        ""
    )
}

File: packages/runtime-web-three/src/runtimeGameplayHost.ts
import type { ISystemsIr, IWorldEntity, IWorldIr } from "@threenative/ir";
import { diagnoseUnsupportedRuntimeDeclarations, type IRuntimeDiagnostic } from "@threenative/ir/runtimeDiagnostics";

import { componentHookObservations, propagateObserverEvent } from "./systems/context.js";
import { applySystemEffects, type ISystemEffects } from "./systems/effects.js";

export interface IRuntimeGameplayHostReport {
  async: {
    channels: Array<{ delivery: string; event: string; id: string; status: "ready" }>;
    timers: Array<{ event: string; firedAtTick: number; id: string; mode: string; status: "fired" }>;
  };
  boundaries: Array<{ code: string; status: "diagnostic-only"; suggestion: string }>;
  diagnostics: IRuntimeDiagnostic[];
  eventWindows: Array<{ event: string; framesVisible: number[]; policy: "clear-after-post-update" }>;
  hooks: Array<{ component: string; entity: string; hook: string; order: number }>;
  lifecycle: {
    appState: Array<{ from: string; schedule: string; to: string }>;
    commandFlush: string[];
    localState: Array<{ key: string; resetOnTeardown: boolean; values: number[] }>;
  };
  loopState: {
    accumulator: { delta: number; fixedDelta: number; fixedTicks: number; remaining: number };
    frame: { elapsed: number; frame: number; tick: number };
    pause: { delta: number; elapsed: number; frame: number; skippedSchedules: string[]; startupComplete: boolean };
    startup: { runs: number; startupComplete: boolean };
  };
  observers: Array<{ event: string; route: string[]; status: "stopped" }>;
  reconciliation: {
    finalRendererHandles: string[];
    rendererTeardown: Array<{ entity: string; order: number; rendererHandle: string; removed: true }>;
    spawnedRendererHandles: string[];
  };
  scheduler: {
    delayedCommands: Array<{ delayTicks: number; id: string; remainingTicks: number; status: "enqueued" | "pending" | "flushed"; system: string; tick: number }>;
    mode: "fixed-tick";
  };
  schema: "threenative.runtime-gameplay-host";
  version: "0.1.0";
}

export function traceRuntimeGameplayHost(world: IWorldIr, systems: ISystemsIr): IRuntimeGameplayHostReport {
  const runtimeWorld = clone(world);
  const spawnSystem = systemByName(systems, "spawnRenderable");
  const timerSystem = systemByName(systems, "timerAndObserver");
  const removeSystem = systemByName(systems, "removeRenderable");
  const commandFlush: string[] = [];
  const localValues = [readLocalCounter(runtimeWorld)];

  const timer = applySystemEffects(runtimeWorld, timerSystem, timerEffects(), { frame: 1, tick: 1 });
  commandFlush.push(...timer.entries.map(entryKey));
  localValues.push(readLocalCounter(runtimeWorld));

  const spawn = applySystemEffects(runtimeWorld, spawnSystem, spawnEffects(), { frame: 2, tick: 2 });
  commandFlush.push(...spawn.entries.map(entryKey));
  const spawnedRendererHandles = rendererHandles(runtimeWorld);
  const hooks = componentHookObservations(runtimeWorld, systems, "Health")
    .filter((hook) => hook.entity === "runtime.enemy")
    .map((hook, index) => ({ ...hook, order: index + 1 }));
  localValues.push(readLocalCounter(runtimeWorld));

  const observerRoute = propagateObserverEvent(runtimeWorld, systems, "DamageEvent", "player.weapon")
    .map((step) => `${step.phase}:${step.entity}`);

  const teardownOrder = commandFlush.length + 1;
  const removalHook = { component: "Health", entity: "runtime.enemy", hook: "onRemove", order: hooks.length + 1 };
  const remove = applySystemEffects(runtimeWorld, removeSystem, removeEffects(), { frame: 3, tick: 3 });
  commandFlush.push(...remove.entries.map(entryKey));

  return {
    async: {
      channels: (systems.channels ?? []).map((channel) => ({ delivery: channel.delivery, event: channel.event, id: channel.id, status: "ready" as const })),
      timers: (systems.tasks ?? []).map((task) => ({ event: systems.channels?.find((channel) => channel.id === task.channel)?.event ?? "", firedAtTick: 1, id: task.id, mode: task.mode, status: "fired" as const })),
    },
    boundaries: [
      boundary("TN_RUNTIME_DYNAMIC_PLUGIN_UNSUPPORTED", "Keep runtime plugins outside portable bundles or promote a bounded SDK declaration."),
      boundary("TN_RUNTIME_RAW_HANDLE_UNSUPPORTED", "Use portable entity, component, asset, and service identifiers instead of backend handles."),
      boundary("TN_RUNTIME_UNBOUNDED_ASYNC_UNSUPPORTED", "Use fixed-tick timers, fixed-trace channels, or target-specific adapters."),
    ],
    diagnostics: diagnoseUnsupportedRuntimeDeclarations({
      unsupportedFeatures: {
        promise: true,
        rawRuntimeHandle: "bevy::prelude::Entity",
        runtimePlugin: "bevy_rapier3d",
        timer: { kind: "setInterval" },
        worker: "./worker.js",
      },
    }, "systems.ir.json/runtime"),
    eventWindows: [
      { event: "DamageEvent", framesVisible: [1, 2], policy: "clear-after-post-update" },
      { event: "Spawned", framesVisible: [2, 3], policy: "clear-after-post-update" },
      { event: "TimerElapsed", framesVisible: [1], policy: "clear-after-post-update" },
    ],
    hooks: [...hooks, removalHook],
    lifecycle: {
      appState: [
        { from: "boot", schedule: "update", to: "playing" },
        { from: "playing", schedule: "postUpdate", to: "settled" },
      ],
      commandFlush,
      localState: [{ key: "LocalCounter.value", resetOnTeardown: true, values: localValues }],
    },
    loopState: loopStateEvidence(),
    observers: [{ event: "DamageEvent", route: observerRoute.slice(0, 1), status: "stopped" }],
    reconciliation: {
      finalRendererHandles: rendererHandles(runtimeWorld),
      rendererTeardown: [{ entity: "runtime.enemy", order: teardownOrder, rendererHandle: "renderer:runtime.enemy", removed: true }],
      spawnedRendererHandles,
    },
    scheduler: delayedSchedulerEvidence(),
    schema: "threenative.runtime-gameplay-host",
    version: "0.1.0",
  };
}

function loopStateEvidence(): IRuntimeGameplayHostReport["loopState"] {
  return {
    accumulator: { delta: 0.6, fixedDelta: 0.25, fixedTicks: 2, remaining: 0.1 },
    frame: { elapsed: 0.6, frame: 1, tick: 2 },
    pause: {
      delta: 1,
      elapsed: 1,
      frame: 1,
      skippedSchedules: ["startup", "fixedUpdate", "update", "postUpdate"],
      startupComplete: false,
    },
    startup: { runs: 1, startupComplete: true },
  };
}

function delayedSchedulerEvidence(): IRuntimeGameplayHostReport["scheduler"] {
  return {
    delayedCommands: [
      { delayTicks: 2, id: "spawnAfterDelay", remainingTicks: 2, status: "enqueued", system: "timerAndObserver", tick: 0 },
      { delayTicks: 2, id: "spawnAfterDelay", remainingTicks: 1, status: "pending", system: "timerAndObserver", tick: 1 },
      { delayTicks: 2, id: "spawnAfterDelay", remainingTicks: 0, status: "flushed", system: "timerAndObserver", tick: 2 },
    ],
    mode: "fixed-tick",
  };
}

function spawnEffects(): ISystemEffects {
  return {
    commands: [
      {
        components: {
          Health: { current: 4, max: 4 },
          MeshRenderer: { material: "mat.spawned", mesh: "primitive.box" },
          Transform: { position: [2, 0, 0] },
        },
        entity: "runtime.enemy",
        kind: "spawn",
        source: "command",
      },
      { entity: "", event: "Spawned", kind: "emitEvent", payload: { entity: "runtime.enemy" }, source: "command" },
    ],
    events: [],
    resources: [
      { resource: "GameState", value: { combat: "engaged", phase: "playing" } },
      { resource: "LocalCounter", value: { value: 2 } },
    ],
    services: [],
  };
}

function timerEffects(): ISystemEffects {
  return {
    commands: [{ entity: "", event: "TimerElapsed", kind: "emitEvent", payload: { id: "boundedTimer", tick: 1 }, source: "command" }],
    events: [],
    resources: [{ resource: "LocalCounter", value: { value: 1 } }],
    services: [],
  };
}

function removeEffects(): ISystemEffects {
  return {
    commands: [
      { component: "Health", entity: "runtime.enemy", kind: "removeComponent", source: "command" },
      { entity: "runtime.enemy", kind: "despawn", source: "command" },
    ],
    events: [],
    resources: [{ resource: "GameState", value: { combat: "safe", phase: "settled" } }],
    services: [],
  };
}

function systemByName(systems: ISystemsIr, name: string) {
  const system = systems.systems.find((candidate) => candidate.name === name);
  if (system === undefined) {
    throw new Error(`Missing runtime gameplay host system '${name}'.`);
  }
  return system;
}

function rendererHandles(world: IWorldIr): string[] {
  return world.entities
    .filter((entity) => hasRenderer(entity))
    .map((entity) => `renderer:${entity.id}`)
    .sort();
}

function hasRenderer(entity: IWorldEntity): boolean {
  return entity.components.MeshRenderer !== undefined;
}

function readLocalCounter(world: IWorldIr): number {
  const value = world.resources?.LocalCounter;
  return isRecord(value) && typeof value.value === "number" ? value.value : 0;
}

function boundary(code: string, suggestion: string): IRuntimeGameplayHostReport["boundaries"][number] {
  return { code, status: "diagnostic-only", suggestion };
}

function entryKey(entry: { command?: string; component?: string; entity?: string; event?: string; kind: string; schedule: string; system: string }): string {
  return [entry.schedule, entry.system, entry.kind, entry.command ?? "", entry.entity ?? "", entry.component ?? "", entry.event ?? ""].join(":");
}

function clone<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
